src/computability/language.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson.
+-/
+import data.finset.basic
+
+/-!
+# Languages
+This file contains the definition and operations on formal languages over an alphabet. Note strings
+are implemented as lists over the alphabet.
+The operations in this file define a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra)
+over the languages.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- A language is a set of strings over an alphabet. -/
+@[derive [has_mem (list α), has_singleton (list α), has_insert (list α)]]
+def language (α) := set (list α)
+
+namespace language
+
+local attribute [reducible] language
+
+instance : has_zero (language α) := ⟨(∅ : set _)⟩
+instance : has_one (language α) := ⟨{[]}⟩
+
+instance : inhabited (language α) := ⟨0⟩
+
+instance : has_add (language α) := ⟨set.union⟩
+instance : has_mul (language α) := ⟨λ l m, (l.prod m).image (λ p, p.1 ++ p.2)⟩
+
+@[simp] lemma zero_def : (0 : language α) = (∅ : set _) := rfl
+@[simp] lemma one_def : (1 : language α) = {[]} := rfl
```

Comment:
Does making these `simp` lemmas mean that the simplifier won't ever try to simplify `1 * a` to `a`, because the `1` gets unfolded first? If so, it might be worth not simplifying.